torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -242,6 +244,72 @@ def aot_dispatch_base(
     return compiled_fn
 
 
+def collect_donated_buffer_metadata_from_tensor(
+    fw_inputs: List[torch.Tensor],
+    fw_outputs: List[torch.Tensor],
+    bw_outputs: List[torch.Tensor],
+    fw_saved_tensor_range: Tuple[int, int],
+) -> List[int]:
+    """
+    Checks if the saved tensors are donated buffers, which means a saved tensor is not
+    an alias of any tensors in fw_inputs, bw_outputs, and fw_outputs (except other saved
+    tensors).
+
+    Layout of fw_outputs:
+        start, end = fw_saved_tensor_range
+        fw_outputs = [*inner_forward_out, *saved_tensors, *saved_symints]
+            where inner_forward_out = fw_outputs[:start]
+                saved_tensors = fw_outputs[start:end]
+                saved_symints = fw_outputs[end:]
+    """
+    fw_saved_tensor_begin, fw_saved_tensor_end = fw_saved_tensor_range
+    num_saved_tensor = fw_saved_tensor_end - fw_saved_tensor_begin
+
+    storage_refs = set()
+    for t in itertools.chain(fw_inputs, fw_outputs[:fw_saved_tensor_begin], bw_outputs):
+        if isinstance(t, torch.Tensor):
+            storage_refs.add(StorageWeakRef(t.untyped_storage()))
+
+    donated_buffer_idxs = []
+    for i in range(num_saved_tensor):
+        t = fw_outputs[fw_saved_tensor_begin + i]
+        if StorageWeakRef(t.untyped_storage()) not in storage_refs:
+            donated_buffer_idxs.append(i)
+
+    return donated_buffer_idxs
+
+
+def collect_donated_buffer_metadata(
+    fw_module: torch.fx.GraphModule,
+    bw_module: torch.fx.GraphModule,
+    num_inner_fwd_outputs: int,
+    num_saved_tensor: int,
+) -> List[int]:
+    """
+    Collects donated buffer metadata from fw_module and bw_module.
+    """
+
+    fw_inputs = fw_module.graph.find_nodes(op="placeholder")
+    bw_outputs = next(reversed(bw_module.graph.find_nodes(op="output"))).args[0]
+    fw_outputs = next(reversed(fw_module.graph.find_nodes(op="output"))).args[0]
+
+    # check if every node has meta["val"] since meta["val"] may be lost during
+    # graph transformation.
+    try:
+        fw_inputs = [n.meta["val"] for n in fw_inputs]
+        fw_outputs = [n.meta["val"] for n in fw_outputs]
+        bw_outputs = [n.meta["val"] for n in bw_outputs]
+    except KeyError:
+        return []
+
+    return collect_donated_buffer_metadata_from_tensor(
+        fw_inputs,
+        fw_outputs,
+        bw_outputs,
+        (num_inner_fwd_outputs, num_inner_fwd_outputs + num_saved_tensor),
```

Comment:
nit: keeping track of the indices of various outputs has historically been pretty hard to reason about - we actually have an existing `CompiledFunction.metadata.tensors_saved_for_backwards_slice` that gives you a start:end range of the indices of graph outs that correspond to save tensors, so I would use that instead of recomputing it here.